presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestTransformCorrelatedScalarSubquery.java

```diff
@@ -0,0 +1,189 @@
+import com.facebook.presto.sql.tree.StringLiteral;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.facebook.presto.sql.tree.WhenClause;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import org.testng.annotations.Test;
+
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.assignUniqueId;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.expression;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.filter;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.lateral;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.markDistinct;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.project;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
+import static com.facebook.presto.sql.planner.iterative.rule.test.PlanBuilder.expressions;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+
+public class TestTransformCorrelatedScalarSubquery
+        extends BaseRuleTest
+{
+    private Rule rule = new TransformCorrelatedScalarSubquery();
+
+    @Test
+    public void doesNotFireOnPlanWithoutLateralNode()
+    {
+        tester().assertThat(rule)
+                .on(p -> p.values(p.symbol("a")))
+                .doesNotFire();
+    }
+
+    @Test
+    public void doesNotFireOnCorrelatedNonScalar()
+    {
+        tester().assertThat(rule)
+                .on(p -> p.lateral(
+                        ImmutableList.of(p.symbol("corr")),
+                        p.values(p.symbol("corr")),
+                        p.values(p.symbol("a"))))
+                .doesNotFire();
+    }
+
+    @Test
+    public void doesNotFireOnUncorrelated()
+    {
+        tester().assertThat(rule)
+                .on(p -> p.lateral(
+                        ImmutableList.<Symbol>of(),
+                        p.values(p.symbol("a")),
+                        p.values(ImmutableList.of(p.symbol("b")), ImmutableList.of(expressions("1")))))
+                .doesNotFire();
+    }
+
+    @Test
+    public void rewritesOnSubqueryWithoutProjection()
+    {
+        tester().assertThat(rule)
+                .on(p -> p.lateral(
+                        ImmutableList.of(p.symbol("corr")),
+                        p.values(p.symbol("corr")),
+                        p.enforceSingleRow(
+                                p.filter(
+                                        p.expression("1 = a"), // TODO use correlated predicate, it requires support for correlated subqueries in plan matchers
+                                        p.values(p.symbol("a"))))))
+                .matches(
+                        project(
+                                filter(
+                                        ensureScalarSubquery(),
+                                        markDistinct(
```

Comment:
we know that `values` produces at most single row here, therefore PROJECT/FILTER/MARK DISTINCT is not needed

Reply:
Yes, but this rule does not check if is subquery is at most scalar. This will be handled before this rule is fired.

Reply:
Ok, I will do it. Thanks!